### PR TITLE
ddl: add primary key default value check when create table

### PR DESCRIPTION
**Pull request by crazycs520**
## What have you changed? (mandatory)
add check for below sqls:
```SQL
mysql > create table t2 (id int default null primary key , age int)
(1067, "Invalid default value for 'id'")
mysql > create table t2 (id int null primary key , age int)
(1171, 'All parts of a PRIMARY KEY must be NOT NULL; if you need NULL in a key, use UNIQUE instead')
mysql > create table t2 (id int default null, age int, primary key(id));
(1171, 'All parts of a PRIMARY KEY must be NOT NULL; if you need NULL in a key, use UNIQUE instead')
mysql > create table t2 (id int null, age int, primary key(id))
(1171, 'All parts of a PRIMARY KEY must be NOT NULL; if you need NULL in a key, use UNIQUE instead')
```

## What is the type of the changes? (mandatory)

- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested? (mandatory)

unit test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Does this PR need to be added to the release notes? (mandatory)
No

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -230,12 +237,12 @@ func setCharsetCollationFlenDecimal(tp *types.FieldType) error {
 }
 
 func buildColumnAndConstraint(ctx sessionctx.Context, offset int,
-	colDef *ast.ColumnDef) (*table.Column, []*ast.Constraint, error) {
+	colDef *ast.ColumnDef, outPriKeyConstraint *ast.Constraint) (*table.Column, []*ast.Constraint, error) {
```
Comment (shenli) [inline ddl/ddl_api.go]: The name of `outPriKeyConstraint ` is not self-explained. Please add a comment for it.

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -263,7 +270,7 @@ func isExplicitTimeStamp() bool {
 }
 
 // columnDefToCol converts ColumnDef to Col and TableConstraints.
-func columnDefToCol(ctx sessionctx.Context, offset int, colDef *ast.ColumnDef) (*table.Column, []*ast.Constraint, error) {
+func columnDefToCol(ctx sessionctx.Context, offset int, colDef *ast.ColumnDef, outPriKeyConstraint *ast.Constraint) (*table.Column, []*ast.Constraint, error) {
```
Comment (shenli) [inline ddl/ddl_api.go]: Please update the comment.

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -347,6 +359,26 @@ func columnDefToCol(ctx sessionctx.Context, offset int, colDef *ast.ColumnDef) (
 				// TODO: Support this type.
 			}
 		}
+
+		// primary key should not be null
```
Comment (winkyao) [inline ddl/ddl_api.go]: Capitalize the first word, and add `.` to the end.

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -347,6 +359,26 @@ func columnDefToCol(ctx sessionctx.Context, offset int, colDef *ast.ColumnDef) (
...
+		if mysql.HasPriKeyFlag(col.Flag) && hasDefaultValue && col.DefaultValue == nil {
+			return nil, nil, types.ErrInvalidDefault.GenByArgs(col.Name)
+		}
+		// set primary key flag for outer primary key constraint.
```
Comment (winkyao) [inline ddl/ddl_api.go]: ditto

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -347,6 +359,26 @@ func columnDefToCol(ctx sessionctx.Context, offset int, colDef *ast.ColumnDef) (
...
+		// such as: create table t1 (id int , age int, primary key(id))
+		if !mysql.HasPriKeyFlag(col.Flag) && outPriKeyConstraint != nil && outPriKeyConstraint.Tp == ast.ConstraintPrimaryKey {
```
Comment (winkyao) [inline ddl/ddl_api.go]: The judgement `outPriKeyConstraint.Tp == ast.ConstraintPrimaryKey` is redundant.
`outPriKeyConstraint != nil` is enough.

File: ddl/ddl_api.go [hunk 6]
```diff
@@ -347,6 +359,26 @@ func columnDefToCol(ctx sessionctx.Context, offset int, colDef *ast.ColumnDef) (
...
+			for _, key := range outPriKeyConstraint.Keys {
+				if key.Column.Name.L != col.Name.L {
+					continue
+				}
+				col.Flag |= mysql.PriKeyFlag
```
Comment (winkyao) [inline ddl/ddl_api.go]: we do not set the PriKeyFlag in this case before?

Comment (crazycs520) [reply ddl/ddl_api.go]: For outer primary key constraint, do not set before.

File: ddl/ddl_api.go [hunk 7]
```diff
@@ -347,6 +359,26 @@ func columnDefToCol(ctx sessionctx.Context, offset int, colDef *ast.ColumnDef) (
 				// TODO: Support this type.
 			}
 		}
+
```
Comment (jackysp) [inline ddl/ddl_api.go]: It's better to use a standalone function to do the primary key check. `columnDefToCol` is too long.

File: ddl/ddl_api.go [hunk 8]
```diff
@@ -173,8 +173,16 @@ func buildColumnsAndConstraints(ctx sessionctx.Context, colDefs []*ast.ColumnDef
 	constraints []*ast.Constraint) ([]*table.Column, []*ast.Constraint, error) {
 	var cols []*table.Column
 	colMap := map[string]*table.Column{}
+	// outPriKeyConstraint is the primary key constraint out of column definition. such as: create table t1 (id int , age int, primary key(id));
+	var outPriKeyConstraint *ast.Constraint
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: This can be moved into `buildColumnAndConstraint` function

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**